gke: add resource quotas for Cilium Namespace

aanm opened the pull request:
When deploying Cilium in its own namespace, it's required to define
resource quotas. For now we will create a ResourceQuota for 10k pods
that are node-critical and 15 pods that are cluster-critical.

Also, add checkers in helm if users try to install Cilium in `kube-system`
namespace with the deprecated labels.

Fixes https://github.com/cilium/cilium/issues/13852

```release-note
Add Resource Quotas in Cilium Namespace for GKE installations
```

aanm committed:
install/kubernetes: warn user for invalid helm options

In GKE environments, if Cilium is deployed in 'kube-system' namespace
with the 'kubernetes.io/cluster-service: "true"' labels, GKE will delete
the DaemonSet within 1 minutes [1]. To avoid this problem these labels
are no longer installed for new installations, however, if the user
tries to install this combination of options, we should fail the
installation and warn the user about the correct usage.

[1] https://github.com/kubernetes/kubernetes/issues/51376
Signed-off-by: André Martins <andre@cilium.io>

pchaigno commented on Documentation/gettingstarted/k8s-install-gke.rst:
```diff
@@ -85,7 +85,7 @@ below. This will ensure all pods are managed by Cilium.
 
     kubectl create namespace cilium
```
This is not needed anymore.

pchaigno commented on test/helpers/cons.go:
```diff
@@ -168,7 +168,7 @@ const (
 	KubeSystemNamespace    = "kube-system"
 	CiliumNamespaceDefault = KubeSystemNamespace
-	CiliumNamespaceGKE     = "cilium"
+	CiliumNamespaceGKE     = KubeSystemNamespace
```
This variable can be removed and `GetCiliumNamespace()` can be simplified or maybe even removed.

pchaigno reviewed (changes requested):
We still create the cilium namespace in `test/gke/select-cluster.sh`. Not sure if we can get rid of that yet.

aanm committed:
gke: add resource quotas for Cilium namespace

When deploying Cilium in its own namespace, it's required to define
resource quotas. For now we will create a ResourceQuota for 10k pods
that are node-critical and 15 pods that are cluster-critical.

Signed-off-by: André Martins <andre@cilium.io>

pchaigno reviewed (approved):
:+1: only for my codeowners (docs-structure) since I don't think I'm qualified to review the rest.

aditighag commented on install/kubernetes/cilium/templates/cilium-resource-quota.yaml:
```diff
@@ -0,0 +1,33 @@
+spec:
+  hard:
+    # 5k nodes * 2 DaemonSets (Cilium and cilium node init)
+    pods: "10k"
```
Just out of curiosity, where can I find more details about this number - 5k nodes from scalability perspective?

aanm replied:
@aditighag https://kubernetes.io/docs/setup/best-practices/cluster-large/#support

joestringer replied:
Surely we should allow users to customize this via Helm?

errordeveloper commented on install/kubernetes/cilium/templates/cilium-resource-quota.yaml:
```diff
@@ -0,0 +1,33 @@
+{{- if and (ne .Release.Namespace "kube-system" ) .Values.gke.enabled }}
```
Don't you think it would be more useful if this was `if and (ne .Release.Namespace "kube-system" )`? ..or perhaps `if and (ne .Release.Namespace "kube-system" ) (resourceQuotas.enabled)`?

errordeveloper replied:
E.g. it could be handy to add include this for OpenShift, where `cilium` namespace is also used.

joestringer commented on install/kubernetes/cilium/templates/cilium-resource-quota.yaml:
```diff
@@ -0,0 +1,33 @@
+spec:
+  hard:
+    # 15 "clusterwide" Cilium Operator pods for HA
+    pods: "15"
```
ditto here?

jrfastab replied:
Don't we want this in this PR so we don't momentarily break someone.

joestringer reviewed (approved):
Minor followup item, I don't expect many users asking for this since the limits should suffice for most scenarios but it's always possible.
